### DISCUSSION TITLE
`useQuery` data return type is `Data | undefined` when using Suspense

### DISCUSSION THREAD

**Discussion by PepijnSenders** [TanStack/query#654] [General]
**Describe the bug**
Not sure if it's possible to fix this, but when we use suspense `{ data } = useQuery()` will never be `undefined`. The return type of `data` is `Data | undefined` though. Maybe we can create another type for when it's suspended? Not sure.

**Comment by jelteliekens** [comment]
Hi, I was wondering this myself and I think it could be possible with discriminated unions where one config type has suspense enabled and another one with suspense disabled. 

The only difficult part is that suspense can be enabled globally but we don't know this in the hooks. This means that you'll need to manually set the correct config type in the hooks which you'll probably don't want. It would be better that they changed automatically of course. Maybe there is an option to let the developer override some types in a declaration file so they only need to change it in 1 place?

I also think there is a similar case when `useErrorBoundary` is enabled. When it is enabled `data` should be just `Data` if the status is not idle or loading. There is no need to check for error since it's being catched by the error boundary.

**Comment by PepijnSenders** [comment]
Yes, exactly! The return type is much simpler when you have both suspense and error boundary enabled.

**Comment by nielsdB97** [comment]
I personally also think that this would be much nicer and would help clean up code in child components even more, like Suspense mode already does so well.

However I read in some issue that this was not feasible because of the fact that  a query can be disabled by setting `enabled: false` in the options. Therefore it could still return `undefined`.
Edit: found [the issue](https://github.com/tannerlinsley/react-query/issues/1297#issuecomment-730462963)
